ores/wsgi/util.py

```diff
@@ -35,23 +37,147 @@ def read_bar_split_param(request, param, default=None, type=str):
+    include_features = 'features' in request.args
+    injection_caches = parse_injection(request, rev_id)
+    model_info = parse_model_info(request)
+
+    if context_name and context_name in scoring_system and not model_names:
+        model_names = scoring_system[context_name].keys()
+
+    return ScoreRequest(context_name, rev_ids, model_names,
+                        precache=precache,
+                        include_features=include_features,
+                        injection_caches=injection_caches,
+                        model_info=model_info)
+
+
+def parse_rev_ids(request, rev_id):
+    if rev_id is not None:
+        return [int(rev_id)]
+    else:
+        return read_bar_split_param(request, "revids", type=int)
+
+
+def parse_model_names(request, model_name):
+    if model_name is not None:
+        return [model_name]
+    else:
+        return read_bar_split_param(request, "models", type=str)
+
+
 def parse_injection(request, rev_id):
     """Parse values for features / datasources of interest."""
     cache = {}
-    try:
-        if 'inject' in request.values:
+
+    if 'inject' in request.values:
+        try:
             cache = json.loads(request.values['inject'])
+        except json.JSONDecodeError as e:
+            raise CacheParsingError(e)
 
-        for k, v in request.values.items():
-            if k.startswith(("feature.", "datasource.")):
-                cache[k] = json.loads(v)
+    if rev_id is not None:
+        rev_cache = cache
 
-        return {rev_id: cache} if len(cache) > 0 else None
-    except Exception as e:
-        raise CacheParsingError(e)
+        try:
+            for k, v in request.values.items():
+                if k.startswith(("feature.", "datasource.")):
+                    rev_cache[k] = json.loads(v)
+        except json.JSONDecodeError as e:
+            raise CacheParsingError(e)
 
+        if len(rev_cache) > 0:
+            cache = {rev_id: rev_cache}
+    else:
+        cache = {int(rev_id): c for rev_id, c in cache.items()}
 
-def jsonify(doc):
-    minify = request.args.get('format') == 'json'
-    separators = None if minify else "  "
-    return jsonify_(doc, separators=separators)
+    return cache or None
+
+
+def parse_model_info(request):
+    return read_bar_split_param(request, "model_info", type=str)
+
+
+def build_score_request_from_event(precache_map, event):
+    context_name = event['database']
+    rev_id = event['rev_id']
+
+    # Check to see if we have the context available in our precache_map
+    if context_name not in precache_map:
+        return None
+
+    # Start building the response document
+    event_set = build_event_set(event)
+
+    model_names = {precache_map[context_name][e] for e in event_set
+                   if e in precache_map[context_name]}
+
+    if len(model_names) == 0:
+        return None
+
+    return ScoreRequest(context_name, model_names, rev_id)
+
+
+# TODO: This strategy for building up events is not sustainable.
```

Comment:
Not exactly.  I think we build up a set of characteristics for each edit and a set of matching characteristics for each event.  In turn, we'll need to allow for configuration to require and exclude certain types of events.  e.g. `{'edit', '!bot', '!new'}` would only match non-page-creating non-bot edits.  `{'log', 'newuser', '!autocreate'}` would only match new user creations that create a new centralauth record.